js/modules/k6/experimental/tracing/trace_id.go

```diff
@@ -22,29 +22,78 @@ const (
 	metadataTraceIDKeyName = "trace_id"
 )
 
-// TraceID represents a trace-id as defined by the [W3c specification], and
+// NewTraceID generates a new trace ID.
+//
+// Because the trace ID relies on randomness, and its underlying serialization
+// depends on the time, it is mandatory to use this function to generate
+// trace IDs, instead of creating them manually.
+//
+// As a result this constructor returns an instantiated TraceIDEncoder, which
+// can be used to encode the trace ID into a string.
+func NewTraceID(prefix int16, code int8, t time.Time, randSource io.Reader) (TraceIDEncoder, error) {
```

Comment:
I would discourage this approach. I think it's not really necessary to limit potential future extensions. We can just use the concrete type as an implementation of TraceIDEncoder.

This is still a very valuable guideline to me. https://github.com/golang/go/wiki/CodeReviewComments#interfaces
>Go interfaces generally **belong in the package that uses values** of the interface type, not the package that implements those values. **The implementing package should return concrete** (usually pointer or struct) types: that way, **new methods can be added to implementations without requiring** extensive **refactoring**.



```suggestion
func NewTraceID(prefix int16, code int8, t time.Time, randSource io.Reader) (TraceID, error) {
```

Reply:
>The problem this tries to solve is that instantiating a traceId directly

This is not a problem generally we should care in Go. We drive this by documentation if it is really required. The standard library has a lot of these cases. (e.g. [File](https://pkg.go.dev/os@go1.20.3#File), you don't expect to use it by doing `File{}`, [ServeMux](https://pkg.go.dev/net/http#ServeMux)).

>This is why the initial implementation did that in the Encode method.

I'm not against it, it think it is good.

>as a public constructor returning a private struct would also feel wrong

It seems that it is used from the same package so in the end, we don't need to export this function. We can have `newTraceID() traceID`.
Instead, if we need TraceID exported for any reason, then we can just have `NewTraceID() TraceID`. All the fields of TraceID are not exported so they are "protected" from external usage.

Reply:
> The standard library has a lot of these cases. (e.g. [File](https://pkg.go.dev/os@go1.20.3#File), you don't expect to use it by doing File{}, [ServeMux](https://pkg.go.dev/net/http#ServeMux)).

This is a fair point indeed. I did not consider it 👍🏻 

> I'm not against it, I think it is good.
@mstoykov disagrees, hence this PR. I'm somewhere in the middle, I must admit, I'm not necessarily horrified by it, but I prefer to have methods that do a single thing, whereas `Encode` before that PR does both serialisation and encoding.

> Instead, if we need TraceID exported for any reason, then we can just have NewTraceID() TraceID. All the fields of TraceID are not exported so they are "protected" from external usage.

I like the middle-ground there indeed 👍🏻 

Consider me convinced, thanks for your input @codebien 🙇🏻 much appreciated!

So in summary, and based on my assumption that @mstoykov would agree, I would keep make everything private, and have the constructor return a `traceId` instance instead of an interface (which I would drop).